### PR TITLE
fix(dht): use new DHKE shared secret type

### PR DESCRIPTION
**Pull request by AaronFeickert**
Description
---
Ensures safer use of ECDH shared secrets by switching to the new `DiffieHellmanSharedSecret` type. Updates `tari-crypto` to v0.15.7 to accomplish this.

Motivation and Context
---
Currently, an ECDH secret used for message keys is produced as a `RistrettoPublicKey`, converted to bytes, and returned as a byte array. However, neither the `RistrettoPublicKey` nor the byte array are cleared when dropped. In conjunction with `tari-crypto` [PR 137](https://github.com/tari-project/tari-crypto/pull/137), this work ensures both the `RistrettoPublicKey` and byte array representations of the ECDH secret are zeroized on drop by using that PR's new `DiffieHellmanSharedSecret` type.

How Has This Been Tested?
---
Tested after applying `tari-crypto` [PR 137](https://github.com/tari-project/tari-crypto/pull/137), which adds the new `DiffieHellmanSharedSecret` generic type.

### CODE REVIEW

File: comms/dht/src/crypt.rs
Comment (sdbondi) [inline comms/dht/src/crypt.rs]: Great, I think that the best choice we have is to heap-allocate the secret as you have, as stack types (like `[u8; 32]`) have a leaking problem as they are moved. [As stated by zeroize](https://docs.rs/zeroize/0.7.0/zeroize/index.html#stackheap-zeroing-notes).

I think we'll have to wait for  https://github.com/tari-project/tari-crypto/pull/137 to be merged so that we can call `k.zeroize()` before returning. Otherwise, `k` would still leak once dropped.

Comment (AaronFeickert) [reply comms/dht/src/crypt.rs]: Yep, merging should be considered blocked until the `tari-crypto` PR is merged.

We shouldn't need to explicitly call `k.zeroize()` anymore, since a recent change to the `tari-crypto` PR returns the shared secret in a `Zeroize` wrapper as well.

Comment (AaronFeickert) [reply comms/dht/src/crypt.rs]: Another thought I had was to build a more comprehensive generic shared secret type that zeroizes on drop and provides its own byte representation in `tari-crypto`. This would allow things like KDFs to do things like `shared_secret.as_bytes()` and mean we wouldn't need to handle the separate `Zeroizing` wrapper around the byte array, increasing safety.

Comment (sdbondi) [reply comms/dht/src/crypt.rs]: Ah yup that's right, missed that the shard_secret function now returns the wrapper. 
I agree it's a bit leaky and because it is that it is now part of the public API of tari-crypto, in most cases, users would have to include zerioze explicitly as a dependency just to use tari-crypto, or you would have to re-export it from tari-crypto.
I don't think that is the end of the world, but a newtype feels more ergonomic and safer here.

Comment (AaronFeickert) [reply comms/dht/src/crypt.rs]: Separate PRs, or part of this and the `tari-crypto` PRs?

**Review by sdbondi — DISMISSED** [review]

**Comment by sdbondi** [comment]
ACK

File: comms/dht/src/crypt.rs
Comment (sdbondi) [reply comms/dht/src/crypt.rs]: Up to you, this can go in as is. If you'd like to continue working in this PR mark as draft and it wont auto-merge :)

Comment (AaronFeickert) [reply comms/dht/src/crypt.rs]: Working on adding this now. It'll be a generic `DiffieHellmanSharedSecret<P>` for arbitrary `Zeroize` public key types, will zeroize on drop, and only provides `as_bytes(&self) -> &[u8]` as visibility into the shared secret. So the caller would need to go out of its way to leave a non-zeroizing copy in memory.

Comment (AaronFeickert) [reply comms/dht/src/crypt.rs]: Done.

**Review by stringhandler — DISMISSED** [review]

**Comment by AaronFeickert** [comment]
Do not merge until `tari-crypto` is tagged and `tari` is updated to reflect the new dependency.

**Comment by hansieodendaal** [comment]
utACK

**Comment by AaronFeickert** [comment]
Now updates `tari-crypto` to v0.15.7 and is safe to merge once CI passes.

**Comment by AaronFeickert** [comment]
This is currently blocked on `tari-crypto` [issue 142](https://github.com/tari-project/tari-crypto/issues/142), which addresses the underlying CI failure.

**Comment by sdbondi** [comment]
@AaronFeickert Try adding the tari_crypto/wasm feature

`base_layer/key_manager/Cargo.toml`

```diff
- wasm = ["wasm-bindgen", "js", "tari_common_types", "console_error_panic_hook"]
+ wasm = ["tari_crypto/wasm", "wasm-bindgen", "js", "tari_common_types", "console_error_panic_hook"]
```

**Comment by stringhandler** [comment]
I had confused this with another issue. This is not network breaking

**Review by sdbondi — APPROVED** [review]

**Review by stringhandler — APPROVED** [review]

**Pull request merged**
